tests/test_rocalutionSolver.cpp

```diff
@@ -0,0 +1,180 @@
+
+template <int bz>
+void readLinearSystem(const std::string& matrix_filename, const std::string& rhs_filename, Matrix<bz>& matrix, Vector<bz>& rhs)
+{
+    {
+        std::ifstream mfile(matrix_filename);
+        if (!mfile) {
+            throw std::runtime_error("Could not read matrix file");
+        }
+        readMatrixMarket(matrix, mfile);
+    }
+    {
+        std::ifstream rhsfile(rhs_filename);
+        if (!rhsfile) {
+            throw std::runtime_error("Could not read rhs file");
+        }
+        readMatrixMarket(rhs, rhsfile);
+    }
+}
+
+template <int bz>
+Dune::BlockVector<Dune::FieldVector<double, bz>>
+getDuneSolution(Matrix<bz>& matrix, Vector<bz>& rhs)
+{
+    Dune::InverseOperatorResult result;
+
+    Vector<bz> x(rhs.size());
+
+    typedef Dune::MatrixAdapter<Matrix<bz>,Vector<bz>,Vector<bz> > Operator;
+    Operator fop(matrix);
+    double relaxation = 0.9;
+    Dune::SeqILU<Matrix<bz>,Vector<bz>,Vector<bz> > prec(matrix, relaxation);
+    double reduction = 1e-2;
+    int maxit = 10;
+    int verbosity = 0;
+    Dune::BiCGSTABSolver<Vector<bz> > solver(fop, prec, reduction, maxit, verbosity);
+    solver.apply(x, rhs, result);
+    return x;
+}
+
+template <int bz>
+Dune::BlockVector<Dune::FieldVector<double, bz>>
+testRocalutionSolver(const boost::property_tree::ptree& prm, Matrix<bz>& matrix, Vector<bz>& rhs)
+{
+    const int linear_solver_verbosity = prm.get<int>("verbosity");
+    const int maxit = prm.get<int>("maxiter");
+    const double tolerance = prm.get<double>("tol");
+    const bool opencl_ilu_parallel(true);
+    const int platformID = 0;
+    const int deviceID = 0;
+    const std::string accelerator_mode("rocalution");
+    const std::string fpga_bitstream("empty");    // unused
+    const std::string linsolver("ilu0");
+    Dune::InverseOperatorResult result;
+
+    Vector<bz> x(rhs.size());
+    auto wellContribs = Opm::WellContributions::create(accelerator_mode, true);
+    std::unique_ptr<Opm::BdaBridge<Matrix<bz>, Vector<bz>, bz> > bridge;
+    try {
+        bridge = std::make_unique<Opm::BdaBridge<Matrix<bz>, Vector<bz>, bz> >(accelerator_mode,
+                                                                               fpga_bitstream,
+                                                                               linear_solver_verbosity,
+                                                                               maxit,
+                                                                               tolerance,
+                                                                               platformID,
+                                                                               deviceID,
+                                                                               opencl_ilu_parallel,
+                                                                               linsolver);
+    } catch (const std::logic_error& error) {
+        BOOST_WARN_MESSAGE(true, error.what());
```

Comment:
Mmh, I don't think we should continue in this case, should we? Maybe this rather calls for `BOOST_REQUIRE` or similar that stops the program?

Reply:
Good point. This is used in the other tests to catch 'device not found' exceptions. These occur when the machine has compile tools for GPUs but no actual GPU to run the test on.
For rocalution, the test checks with `rocalution_backend_descriptor->accelerator` if an accelerator is connected. If not, the test is skipped.
We could rethrow the error here, but since no error is ever handled in the catch, we could also just remove the whole `try` block.

Reply:
Ah, i see. The test should work without devices, I guess. You choose the proper solution.